File: algorithms/c#/QuickSort.cs
using System;

namespace Algorithms
{
    class Quicksort
    {
        /// <summary>
        /// Takes the last element as pivot and places that element into its correct location in sorted array.
        /// Places all smaller (than pivot) values to left of pivot and greater (than pivot) values to the right of pivot
        /// </summary>
        /// <param name="arr"></param>
        /// <param name="low"></param>
        /// <param name="high"></param>
        /// <returns></returns>
        static int partition(int[] arr, int low, int high)
        {
            int piv = arr[high];

            // index of smaller element
            int i = (low - 1);
            for (int j = low; j < high; j++)
            {
                // if the current element is smaller than the pivot
                if (arr[j] < piv)
                {
                    i++;

                    // swap arr[i] and arr[j]
                    int loopTemp = arr[i];
                    arr[i] = arr[j];
                    arr[j] = loopTemp;
                }
            }

            // swap arr[i+1] and arr[high] (or pivot)
            int temp = arr[i + 1];
            arr[i + 1] = arr[high];
            arr[high] = temp;

            return (i + 1);
        }

        static void quickSort(int[] arr, int low, int high)
        {
            if (low < high)
            {
                // pi is partitioning index, arr[pi] is now at right place
                int pi = partition(arr, low, high);

                // recursively sort elements before and after partition
                quickSort(arr, low, pi - 1);
                quickSort(arr, pi + 1, high);
            }
        }

        static void printArray(int[] arr)
        {
            for (int i = 0; i < arr.Length; i++)
                Console.WriteLine(arr[i] + " ");
        }

        static void Main(string[] args)
        {
            int[] arr = { 10, 7, 8, 9, 1, 5 };

            Console.WriteLine("Original array: ");
            printArray(arr);

            quickSort(arr, 0, (arr.Length - 1));
            
            Console.WriteLine("\nSorted array: ");
            printArray(arr);
        }
    }
}

File: algorithms/python/quicksort.py
# This function takes the last element as pivot, places the
# pivot element into the correct position in the sorted array,
# then places all smaller (than pivot) elements to the left of
#  pivot and all larger (than pivot) elements to the right of pivot


def partition(arr, low, high):
    i = (low - 1)  # index of smaller element
    pivot = arr[high]  # pivot

    for j in range(low, high):
        # if the current element is smaller than the pivot
        if arr[j] < pivot:
            # increment index of smaller element
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


# the main function that implements quickSort
# arr[] --> Array to be sorted
# low   --> Start index
# high  --> End index


def quicksort(arr, low, high):
    if low < high:
        # pi is partitioning index, arr[pi] is now
        # at the right place
        pi = partition(arr, low, high)

        # Separately sort elements before
        # partition and after partition
        quicksort(arr, low, pi - 1)
        quicksort(arr, pi + 1, high)


# Driver code for test
arr = [10, 7, 8, 9, 1, 5]
n = len(arr)
print("Original array: ")
for i in range(n):
    print("%d " % arr[i])
quicksort(arr, 0, n-1)
print("Sorted array: ")
for i in range(n):
    print("%d " % arr[i])
